File: ui/src/utils/validation.ts
export interface ValidationResult {
    isValid: boolean;
    error?: string;
}

export const validateUsername = (username: string): ValidationResult => {
    if (!username) {
        return { isValid: false, error: "Username is required" };
    }

    if (username.length < 3) {
        return {
            isValid: false,
            error: "Username must be at least 3 characters long",
        };
    }

    if (username.length > 20) {
        return {
            isValid: false,
            error: "Username must be less than 20 characters long",
        };
    }

    if (!/^[a-zA-Z0-9_-]+$/.test(username)) {
        return {
            isValid: false,
            error: "Username can only contain letters, numbers, underscores, and hyphens",
        };
    }

    return { isValid: true };
};

export const validateMessage = (message: string): ValidationResult => {
    if (!message) {
        return { isValid: false, error: "Message is required" };
    }

    if (message.length > 1000) {
        return {
            isValid: false,
            error: "Message must be less than 1000 characters long",
        };
    }

    // Check for potentially harmful content
    const harmfulPatterns = [
        /<script\b[^<]*(?:(?!<\/script>)<[^<]*)*<\/script>/gi, // Script tags
        /javascript:/gi, // JavaScript protocol
        /data:/gi, // Data protocol
        /vbscript:/gi, // VBScript protocol
    ];

    for (const pattern of harmfulPatterns) {
        if (pattern.test(message)) {
            return {
                isValid: false,
                error: "Message contains potentially harmful content",
            };
        }
    }

    return { isValid: true };
};

File: server/internal/validation/validation.go
package validation

import (
	"regexp"
	"strings"
)

type ValidationError struct {
	Field   string
	Message string
}

func (e *ValidationError) Error() string {
	return e.Message
}

func ValidateUsername(username string) error {
	if strings.TrimSpace(username) == "" {
		return &ValidationError{
			Field:   "username",
			Message: "Username is required",
		}
	}

	if len(username) < 3 {
		return &ValidationError{
			Field:   "username",
			Message: "Username must be at least 3 characters long",
		}
	}

	if len(username) > 20 {
		return &ValidationError{
			Field:   "username",
			Message: "Username must be less than 20 characters long",
		}
	}

	validUsername := regexp.MustCompile(`^[a-zA-Z0-9_-]+$`)
	if !validUsername.MatchString(username) {
		return &ValidationError{
			Field:   "username",
			Message: "Username can only contain letters, numbers, underscores, and hyphens",
		}
	}

	return nil
}

func ValidateMessage(message string) error {
	if strings.TrimSpace(message) == "" {
		return &ValidationError{
			Field:   "message",
			Message: "Message is required",
		}
	}

	if len(message) > 1000 {
		return &ValidationError{
			Field:   "message",
			Message: "Message must be less than 1000 characters long",
		}
	}

	// Check for potentially harmful content
	harmfulPatterns := []string{
		`<script\b[^<]*(?:(?!<\/script>)<[^<]*)*<\/script>`, // Script tags
		`javascript:`, // JavaScript protocol
		`data:`,      // Data protocol
		`vbscript:`,  // VBScript protocol
	}

	for _, pattern := range harmfulPatterns {
		matched, err := regexp.MatchString(pattern, message)
		if err != nil {
			continue
		}
		if matched {
			return &ValidationError{
				Field:   "message",
				Message: "Message contains potentially harmful content",
			}
		}
	}

	return nil
} 